### PR TITLE
Task 3.2

### PR DESCRIPTION
**Pull request by shigontsev**

### CODE REVIEW

File: Task_3/Task_3.2/MyDLL/CycledDynamicArray.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        {            
+            while (true)
+            {
+                IEnumerator a = base.GetEnumerator();
+                int i = 0;
+                while (a.MoveNext() && i < Length)
+                {
+                    yield return (T)a.Current;
+                    i++;
+                }
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/CycledDynamicArray.cs]: Using foreach would save you complexity

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace MyDLL
+{
+    public class DynamicArray<T>: IEnumerable<T>, IEnumerable, ICloneable
+    {
+        private const int size = 8;
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: If it's a constant name it right - it's `DefaultCapacity` (PascalCase)

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            mainArray = new T[capacity];
+        }
+
+        public DynamicArray(IEnumerable<T> sourse) : this()
+        {
+            T[] tempArr = sourse.ToArray();
+            ReSize(tempArr.Length);
+
+            Array.Copy(tempArr, mainArray, tempArr.Length);
+        }
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: Could save memory allocation (and processor time btw) just by creating array of needed capacity. Or even setting mainArray to sourse.ToArray() and Length to its length

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+        public void Add(T item)
+        {
+            if (Length+1 >= Capacity)
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]:
```suggestion
            if (Length = Capacity)
```
`[1,2,3,4,5]` - capacity 5, length 5, next add needs a resized array

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            }
+            ReSize(Length + 1);
+
+            Array.Copy(mainArray, index, mainArray, index + 1, Length - index);
+            mainArray[index] = item;
+            return true;
+        }
+
+
+        private void ReSize(int length)
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: requestedCapacity (confusing naming)

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            if (Capacity < length)
+            {
+                int tempSize = size;
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: I look here and see reading of a field. But in reality `size` is a constant irrelevant to object state

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                while (tempSize < length)
+                {
+                    tempSize *= 2;
+                }
+
+                //which ToArray is better?
+                var tempArr = mainArray.ToArray();
+                //T[] tempArr = this.ToArray();
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: var better
calling to array without mediator is better

array.Clone is even better - no need to enumerate

Array.Copy to a new bigger array and replacing mainArray with it - the best - don't need to clone an array just to copy it afterwards

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                return mainArray[index];
+            }
+            set
+            {
+                index = HelperOfIndex(index);
+                mainArray[index] = value;
+            }
+        }
+
+        private int HelperOfIndex(int index)
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: "Helper" doesn't tell anything - it's a name for a utility class with different "tools"
Better name is `ResolveIndex`

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                return index + Length;
+            }
+        }
+
+        public virtual IEnumerator<T> GetEnumerator()
+        {
+            //for (int i = 0; i < Length; i++)
+            //{
+            //    yield return mainArray[i];
+            //}
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: commented code
(with a easier/better solution though)

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        public T[] ToArray()
+        {
+            T[] newArr = new T[Length];
+            Array.Copy(mainArray, newArr, Length);
+            return newArr;
+        }
+
+        public void EditCapacity(int capacity)
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: SetCapacity

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            if (capacity == this.Capacity)
+            {
+                return;
+            }
+            else if (capacity < 0)
+            {
+                throw new ArgumentOutOfRangeException();
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: No exception message

File: Task_3/Task_3.2/MyDLL/DynamicArray.cs [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            if (Length+1 >= Capacity)
+            {
+                ReSize(Length + 1);
+                mainArray[Length - 1] = item;
+            }
+            else
+            {
+                Length++;
+                mainArray[Length - 1] = item;
```
Comment (STeeL835) [inline Task_3/Task_3.2/MyDLL/DynamicArray.cs]: Code duplication - move outside of if

**Review by STeeL835 — APPROVED** [review]

**Comment by shigontsev** [comment]
end

**Pull request merged**